### PR TITLE
Update KServe 2024-2025 Roadmap

### PR DESCRIPTION
**Pull request by yuzisun**
Update KServe roadmap and add planned GenAI inference roadmap items

### CODE REVIEW

File: ROADMAP.md [hunk 1]
```diff
@@ -19,16 +40,13 @@ Reference: [Control plane issues](https://github.com/kserve/kserve/issues?q=is%3
...
 ## Objective: "Graduate KServe Python SDK to 1.0“
 
-- Improve KServe Python SDK dependency management with Poetry
 - Create standarized model packaging API
 - Improve KServe model server observability with metrics and distruted tracing
 - Support batch inference
 
 Reference：[Python SDK issues](https://github.com/kserve/kserve/issues?q=is%3Aissue+is%3Aopen+label%3Akserve%2Fsdk), [Storage issues](https://github.com/kserve/kserve/issues?q=is%3Aissue+is%3Aopen+label%3Akfserving%2Fstorage)
 
 ## Objective: "Graduate ModelMesh to beta"
```
Comment (yuzisun) [inline ROADMAP.md]: @ckadner Anything to update for ModelMesh?

File: ROADMAP.md [hunk 2]
```diff
@@ -37,7 +55,6 @@ Reference: [ModelMesh issues](https://github.com/kserve/modelmesh-serving/issues
 
 ## Objective: "Graduate InferenceGraph to beta"
```
Comment (yuzisun) [inline ROADMAP.md]: @rachitchauhan43 Anything to update for InferenceGraph?

File: ROADMAP.md [hunk 3]
```diff
@@ -1,16 +1,37 @@
-# KServe 2023 Roadmap
+# KServe 2024-2025 Roadmap
+## Objective: "Support GenAI inference"
```
Comment (yuzisun) [inline ROADMAP.md:2]: @terrytangyuan to review

File: ROADMAP.md (hunk 2 above)
Comment (rachitchauhan43) [reply ROADMAP.md]: Hey @yuzisun, nothing that I could think of apart from making RawDeployment mode more robust.

@bmopuri : Anything on top of your mind ?

File: ROADMAP.md [hunk 4]
```diff
@@ -1,16 +1,37 @@
...
 
 ## Objective: "Graduate core inference capability to stable/GA"
-- Promote `InferenceService` and `ClusterServingRuntime`/`ServingRuntime` CRD from v1beta1 to v1 
+- Promote `InferenceService` and `ClusterServingRuntime`/`ServingRuntime` CRD to v1
   * Improve `InferenceService` CRD for REST/gRPC protocol interface
   * Unify model storage spec and implementation between KServe and ModelMesh
-  * Add Status to `ServingRuntime` for both ModelMesh and KServe, surface `ServingRuntime` validation errors and deployment status
-  * Deprecate `TrainedModel` CRD and use `InferenceService` annotation to allow dynamic model updates as alternative option to storage initializer
-  * Collocate transformer and predictor in the pod to reduce sidecar resources and networking latency
+  * Deprecate `TrainedModel` CRD and `InferenceService` to support multiple models for model collocations, LoRA adapters.
```
Comment (rhuss) [inline ROADMAP.md]: Does this mean that `InferenceService` is planned to become deprecated ? I don't think so, but the wording here is misleading.

File: ROADMAP.md [hunk 5]
```diff
@@ -1,16 +1,37 @@
-# KServe 2023 Roadmap
+# KServe 2024-2025 Roadmap
```
Comment (terrytangyuan) [inline ROADMAP.md:1]: Would it be a good idea to link each item to an existing/new GitHub issue for deeper discussions on each topic?

File: ROADMAP.md [hunk 6]
```diff
@@ -1,16 +1,37 @@
...
+
+- LLM Autoscaling
+   * Support Model Caching with automatic PV/PVC provisioning.
+   * Support Autoscaling settings for serving runtimes.
+   * Support Autoscaling based on custom metrics.
+
+- LLM RAG Pipeline Orchestration
+   * Support declarative RAG with KServe Inference Graph.
+
+- GenAI Task APIs
```
Comment (danielezonca) [inline ROADMAP.md]:
```suggestion
- Open Inference Protocol extension to GenAI Task APIs
```
What about making more explicit that this point is about Open Inference Protocol?

File: ROADMAP.md (hunk 4 above)
Comment (terrytangyuan) [reply ROADMAP.md]: I think the intention here is to _**extend**_  `InferenceService` to support other types of mechanisms and hence it's an item under this "Promote `InferenceService`...CRD to v1" section

@yuzisun to double check

File: ROADMAP.md (hunk 2 above)
Comment (taneem-ibrahim) [reply ROADMAP.md]: I know @israel-hdez has been looking into the IG support. So tagging him for feedback here.

File: ROADMAP.md [hunk 7]
```diff
@@ -1,16 +1,37 @@
...
+   * Support LoRA adapters.
+   * Support LLM Serving runtimes for TensorRT-LLM, TGI and provide benchmarking comparisions.
+   * Support multi-host, multi-GPU inference runtime.
+
+- LLM Autoscaling
+   * Support Model Caching with automatic PV/PVC provisioning.
+   * Support Autoscaling settings for serving runtimes.
+   * Support Autoscaling based on custom metrics.
+
+- LLM RAG Pipeline Orchestration
```
Comment (skonto) [inline ROADMAP.md]: Do we want to consider support for agentic workflows? cc @pierdipi

Comment (Cali0707) [reply ROADMAP.md]: I'd be happy to work with @pierDipi and others interested on supporting agentic workflows here (if that makes sense)

File: ROADMAP.md [hunk 8]
```diff
@@ -1,16 +1,37 @@
...
+   * Support declarative RAG with KServe Inference Graph.
+
+- GenAI Task APIs
+   * Community maintained Open Inference Protocol repo for OpenAI schema.
+   * Support verical GenAI Task APIs such as embedding, Text-to-Image, Text-To-Code, Doc-To-Text.
+
+- LLM Gateway
+   * Support multiple LLM providers.
+   * Support token based rating limiting.
```
Comment (pierDipi) [inline ROADMAP.md]: nit:

```suggestion
   * Support token based rate limiting.
```

File: ROADMAP.md (hunk 7 above)
Comment (pierDipi) [reply ROADMAP.md]: I believe it's better to discuss at an high level what that means in an issue and whether it's in scope for kserve or not before adding it to the roadmap doc?

Comment (pierDipi) [reply ROADMAP.md]: I've opened an issue here https://github.com/kserve/kserve/issues/3829

File: ROADMAP.md [hunk 9]
```diff
@@ -1,16 +1,37 @@
...
+   * Support Autoscaling based on custom metrics.
+
+- LLM RAG Pipeline Orchestration
+   * Support declarative RAG with KServe Inference Graph.
+
+- GenAI Task APIs
+   * Community maintained Open Inference Protocol repo for OpenAI schema.
+   * Support verical GenAI Task APIs such as embedding, Text-to-Image, Text-To-Code, Doc-To-Text.
+
+- LLM Gateway
```
Comment (lizzzcai) [inline ROADMAP.md:21]: Is there a plan to support the Kubernetes Gateway API?

Comment (lizzzcai) [inline ROADMAP.md:21]: Is there a plan to support the Kubernetes Gateway API?

Comment (yuzisun) [reply ROADMAP.md:21]: Yes kserve is going to integrate with envoy gateway

File: ROADMAP.md [hunk 10]
```diff
@@ -1,16 +1,37 @@
...
+   * Support multiple LLM providers.
+   * Support token based rating limiting.
+   * Support LLM router with traffic shaping, fallback, load balancing.
 
 ## Objective: "Graduate core inference capability to stable/GA"
-- Promote `InferenceService` and `ClusterServingRuntime`/`ServingRuntime` CRD from v1beta1 to v1 
+- Promote `InferenceService` and `ClusterServingRuntime`/`ServingRuntime` CRD to v1
   * Improve `InferenceService` CRD for REST/gRPC protocol interface
```
Comment (spolti) [inline ROADMAP.md:29]: Do we have the list of improvements?

File: ROADMAP.md [hunk 11]
```diff
@@ -19,16 +40,13 @@ Reference: [Control plane issues](https://github.com/kserve/kserve/issues?q=is%3
 
 ## Objective: "Graduate KServe Python SDK to 1.0“
 
-- Improve KServe Python SDK dependency management with Poetry
```
Comment (spolti) [inline ROADMAP.md:22]: Is it worth mentioning the improvement around the decoupling of the storage-initializer dependencies from KServe? https://github.com/kserve/kserve/issues/3489

File: ROADMAP.md (hunk 9 above)
Comment (lizzzcai) [reply ROADMAP.md:21]: how about the metering/usage part? I think it is quite an important feature for a gateway.

File: ROADMAP.md (hunk 1 above)
Comment (terrytangyuan) [reply ROADMAP.md]: cc @israel-hdez @spolti

File: ROADMAP.md (hunk 5 above)
Comment (yuzisun) [reply ROADMAP.md:1]: linked the issues for these items

File: ROADMAP.md (hunk 7 above)
Comment (yuzisun) [reply ROADMAP.md]: added to the roadmap, I think this should be in the scope of KServe to orchestrate RAG/Agent workflow

**Review by terrytangyuan — APPROVED** [review]
/lgtm

**Pull request merged**